### PR TITLE
Feat: Add ios support for accessing WebVTT Subtitle Content 

### PR DESCRIPTION
**Pull request by coofzilla**
## 📝 Summary
This PR adds the capability to access and expose the content of subtitles within RN; but, currently only implements the `ios` portion. The core of this functionality leverages `AVFoundation framework's` `AVPlayerItemLegibleOutput`, enabling the app to intercept subtitle data as attributed strings during video playback. This data is subsequently passed to do the RN layer through the newly introduced event, `onSubtitleTracks`; thus, allowing components to reactively update based on this `subtitle` content.

### Motivation

Our motivation for integrating the onSubtitleTracks event handler with `react-native-video` is to enhance user engagement with HLS VODs of sports content. Utilizing WebVTT tracks, we aim to dynamically display game information like scores as overlays on the video, based on the playback position. This feature ensures viewers have access to contextual sports data as they navigate through the video, significantly improving their viewing experience by making it more informative and interactive. This implementation aligns with the capabilities of the WebVTT API, optimizing our app's use of text tracks for an enriched sports viewing experience. 

> The primary purpose of WebVTT files is to add text overlays to a `<video>`

[source](https://developer.mozilla.org/en-US/docs/Web/API/WebVTT_API)

## 📲 iOS Implementation:
- `RCTPlayerObserver` class implements `AVPlayerItemLegibleOutputPushDelegate` to observe and handle subtitle tracks.
- Configured `legibleOutput` on the player item and set the current class as the delegate to manage legible output.
- Implemented the `legibleOutput` delegate method to capture subtitle data, processing it via the `handleLegibleOutput` method to extract and utilize subtitle strings.

## 🌉 RN Bridge:
- Added an @objc property `onSubtitleTracks` to facilitate event emission, enabling the RN layer to receive subtitle data.
- Implemented method `handleLegibleOutput` to handle attributed strings from subtitles, triggering the `onSubtitleTracks` event with the subtitle text for consumption within RN components.

## 🥳 RN Layer
- Exposed the `onSubtitleTracks` property. 

Example:

```ts
const App = () => {
  const handleSubtitleTracks = useCallback((e: OnSubtitleTracksData) => {
    console.log('Subtitles:', e.subtitleTracks);
  }, []);

  return (
    <Video
      source={{ uri: 'video-url.mp4' }} // Replace 'video-url.mp4' with your video source
      onSubtitleTracks={handleSubtitleTracks} // new event handler
      // Other props for the Video component
    />
  );
};
```

## 📄 Docs
- updated `events.md` to include a description aligned with current standards.

## Test plan

✅ Tested in example app with `uri` included in this PR

Set the tracks you want to use, for my demo I used the below; however, `SelectedTrackType.LANGUAGE` also works. For my actual use-case, we'll be using `TITLE` so I wanted to make sure that worked.

  - Apply following changes to test in example app:
  
  ```js
// VideoPlayer.tsx

      // selectedTextTrack={this.state.selectedTextTrack}
      selectedTextTrack={{
        type: SelectedTrackType.TITLE,
        value: 'English',
      }}
```
Comment out this line; due to `textTracks` being undefined, I believe it has to do with how `data.textTracks` is being set; but, I didn't want to address that on this PR as its a separate issue.

```js
// VideoPlayer.tsx
  {/* {this.state.textTracks.map(track => (
    <Picker.Item
      label={track.language}
      value={track.language}
      key={track.language}
    />
  ))} */}
```
Comment out this line for the app to build( I believe this is a known issue):
```js
// metro.config.js

// /(.*\/react-native-video\/node_modules\/.*)$/,
```

Demo of how it works:

https://github.com/react-native-video/react-native-video/assets/75731066/c5ec08dc-567d-4704-a1b1-a6f39121673f

### CODE REVIEW

File: docs/pages/component/events.md
Comment (freeboub) [inline docs/pages/component/events.md]: This Pr looks good to me, I am just not sure about this naming,
onTextTracks event is send when track list is updated, here I would expect the callback is thrown when subtitle track list is updated, but this is the new functionality here.
what about naming it "onTextTrackDataChanged" ?

File: docs/pages/component/events.md
Comment (freeboub) [inline docs/pages/component/events.md]:
```suggestion
Callback function that is called when new subtitle data is available. It provides the actual subtitle content for the current selected text track, if available (mainly WebVTT). 
```

File: docs/pages/component/events.md
Comment (freeboub) [inline docs/pages/component/events.md:509]: do it make sens to add a duration ? not sure about ios api & webVtt format but in terms of api it would make sense (not mandatory to merge this PR)

File: examples/basic/src/VideoPlayer.tsx
Comment (freeboub) [inline examples/basic/src/VideoPlayer.tsx]:
```suggestion
      description: 'sintel with subtitles',
```

File: examples/basic/src/VideoPlayer.tsx
Comment (freeboub) [inline examples/basic/src/VideoPlayer.tsx:241]: thank you 🙏

**Comment by freeboub** [comment]
few comments, but looks good !
on android I think the code shall be in onCue function if you want to implement it.

**Review by freeboub — CHANGES_REQUESTED** [review]

**Comment by coofzilla** [comment]
> few comments, but looks good ! on android I think the code shall be in onCue function if you want to implement it.

Thank you for the feedback 🙏🏽! I'll take a look at the `android` implementation once I get this merged 😄 I'm going to apply the requested changes here shortly

File: docs/pages/component/events.md
Comment (coofzilla) [reply docs/pages/component/events.md:509]: Yes it does, I can take a look at this and add it as another property on another PR 😊

File: docs/pages/component/events.md
Comment (coofzilla) [reply docs/pages/component/events.md]: This makes perfect sense, thank you, I'm about to push these changes. I think it came from a misunderstanding of what `onTextTracks` was; so, thank you for clearing that up 🙌🏽

**Comment by coofzilla** [comment]
All feedback has been addressed :) 

I've also tested it again in the sample app. 
image

**Comment by coofzilla** [comment]
> few comments, but looks good ! on android I think the code shall be in onCue function if you want to implement it.

yep, I can take a look at this and will submit another PR for it.

File: src/Video.tsx
Comment (KrzysztofMoch) [inline src/Video.tsx]: @coofzilla How about to omit `target` property to don't confuse users ?
Thanks for PR!

Comment (coofzilla) [reply src/Video.tsx]: No problem :)

How does this look?

```ts
    const _onTextTrackDataChanged = useCallback(
      (
        e: NativeSyntheticEvent<OnTextTrackDataChangedData & {target?: number}>,
      ) => {
        const {...eventData} = e.nativeEvent;
        delete eventData.target;
        onTextTrackDataChanged?.(eventData as OnTextTrackDataChangedData);
      },
      [onTextTrackDataChanged],
    );
  ```
  
  I used delete because if I did const `{target, ...eventData}` I'd get the eslint about no unused vars.

image

Comment (KrzysztofMoch) [reply src/Video.tsx]: It's good! Thanks

**Review by KrzysztofMoch — APPROVED** [review]
LGTM, Thank you very much 👍
Let's wait for @freeboub review (this may take some time, for what I know he's busy lately)

**Comment by coofzilla** [comment]
> LGTM, Thank you very much 👍 Let's wait for @freeboub review (this may take some time, for what I know he's busy lately)

Awesome, happy to help! Excited to get this merged in 😊

**Review by freeboub — APPROVED** [review]

**Comment by coofzilla** [comment]
@KrzysztofMoch any other blockers I need to take care of to get this merged? 😅

**Comment by KrzysztofMoch** [comment]
No 😄

**Pull request merged**
